### PR TITLE
Mcol 4234 order by group by mishap

### PR DESCRIPTION
**Pull request by mariadb-SergeyZefirov**
A draft PR for others to look at.

### CODE REVIEW

**Comment by mariadb-SergeyZefirov** [comment]
Main idea here is to remove internal error from ha_mcs_execplan and move analyses and checks from joblist-related code to ha_msc_execplan, if possible.

File: dbcon/execplan/calpontselectexecutionplan.cpp [hunk 1]
```diff
@@ -100,12 +100,13 @@ CalpontSelectExecutionPlan::CalpontSelectExecutionPlan(const int location)
 
 CalpontSelectExecutionPlan::CalpontSelectExecutionPlan(
     const ReturnedColumnList& returnedCols, ParseTree* filters, const SelectList& subSelects,
-    const GroupByColumnList& groupByCols, ParseTree* having, const OrderByColumnList& orderByCols,
+    const GroupByColumnList& groupByCols, uint32_t groupByKeysCount, ParseTree* having, const OrderByColumnList& orderByCols,
     const string alias, const int location, const bool dependent, const bool withRollup)
  : fReturnedCols(returnedCols)
  , fFilters(filters)
  , fSubSelects(subSelects)
  , fGroupByCols(groupByCols)
```
Comment (drrtuy) [inline dbcon/execplan/calpontselectexecutionplan.cpp:108]: Is there a difference b/w `groupByCols` and introduced `groupByKeysCount`?

File: dbcon/execplan/calpontselectexecutionplan.h [hunk 2]
```diff
@@ -150,6 +150,7 @@ class CalpontSelectExecutionPlan : public CalpontExecutionPlan
 
   CalpontSelectExecutionPlan(const ReturnedColumnList& returnedCols, ParseTree* filters,
                              const SelectList& subSelects, const GroupByColumnList& groupByCols,
+                             uint32_t groupByKeysCount,
```
Comment (drrtuy) [inline dbcon/execplan/calpontselectexecutionplan.h:153]: This can be constant.

File: dbcon/joblist/limitedorderby.cpp [hunk 3]
```diff
@@ -70,7 +70,7 @@ void LimitedOrderBy::initialize(const RowGroup& rg, const JobInfo& jobInfo, bool
   for (; i != jobInfo.orderByColVec.end(); i++)
   {
     map<uint32_t, uint32_t>::iterator j = keyToIndexMap.find(i->first);
-    idbassert(j != keyToIndexMap.end());
+    //idbassert(j != keyToIndexMap.end());
```
Comment (drrtuy) [inline dbcon/joblist/limitedorderby.cpp:73]: If it is commented then remove it.

File: dbcon/joblist/tupleaggregatestep.cpp [hunk 4]
```diff
@@ -981,6 +981,7 @@ void TupleAggregateStep::prep1PhaseAggregate(JobInfo& jobInfo, vector<RowGroup>&
     groupBy.push_back(groupby);
     groupbyMap.insert(make_pair(jobInfo.groupByColVec[i], i));
   }
+  uint64_t lastActualGbIndex = jobInfo.keysInGbCols;
```
Comment (drrtuy) [inline dbcon/joblist/tupleaggregatestep.cpp:985]: This can be a const.

File: dbcon/joblist/tupleaggregatestep.cpp [hunk 5]
```diff
@@ -1485,6 +1487,7 @@ void TupleAggregateStep::prep1PhaseAggregate(JobInfo& jobInfo, vector<RowGroup>&
 void TupleAggregateStep::prep1PhaseDistinctAggregate(JobInfo& jobInfo, vector<RowGroup>& rowgroups,
                                                      vector<SP_ROWAGG_t>& aggregators)
 {
+  idbassert(jobInfo.groupByColVec.size() == jobInfo.keysInGbCols); // MCOL-4234, see above how to handle. Most pobably, groupByColVec.size() is zero and all keys there are ORDER BY keys.
```
Comment (drrtuy) [inline dbcon/joblist/tupleaggregatestep.cpp:1491]: Put the comment on top b/c a long comment on the same line is less readable.

File: dbcon/mysql/ha_mcs_execplan.cpp [hunk 6]
```diff
@@ -8307,6 +8309,15 @@ int getSelectPlan(gp_walk_info& gwi, SELECT_LEX& select_lex, SCSEP& csep, bool i
           if ((ord_item->type() == Item::CONST_ITEM && ord_item->cmp_type() == INT_RESULT) &&
               ord_item->full_name() && !strcmp(ord_item->full_name(), "Not_used"))
           {
+            // This is the case of server's optimization of EXISTS(SELECT ...).
+            // look at sql_base.cc.
+            continue;
+          }
+          else if (ord_item->type() == Item::NULL_ITEM)
+          {
+            // Do not add NULL contant as a order key.
```
Comment (drrtuy) [inline dbcon/mysql/ha_mcs_execplan.cpp:8356]: Type constant

File: dbcon/mysql/ha_mcs_execplan.cpp [hunk 7]
```diff
@@ -8307,6 +8309,15 @@ int getSelectPlan(gp_walk_info& gwi, SELECT_LEX& select_lex, SCSEP& csep, bool i
...
+            // NULL as order key is an optimization hint for Mysql, but
```
Comment (drrtuy) [inline dbcon/mysql/ha_mcs_execplan.cpp:8357]: MariaDB.

File: dbcon/mysql/ha_mcs_execplan.cpp [hunk 8]
```diff
@@ -8357,7 +8368,12 @@ int getSelectPlan(gp_walk_info& gwi, SELECT_LEX& select_lex, SCSEP& csep, bool i
         else
           rc->asc(false);
 
-        gwi.orderByCols.push_back(SRCP(rc));
+        SRCP srcprc(rc);
+        gwi.orderByCols.push_back(srcprc);
+        if (orderbyOutOfGroupBy[indexOC])
```
Comment (drrtuy) [inline dbcon/mysql/ha_mcs_execplan.cpp:8411]: Is there a guarantee `indexOC` value is less than the size of the `orderbyOutOfGroupBy`?

**Review by drrtuy — CHANGES_REQUESTED** [review]
Plz take a look at the comments.

File: dbcon/mysql/ha_mcs_execplan.cpp (hunk 7 above)
Comment (mariadb-SergeyZefirov) [reply dbcon/mysql/ha_mcs_execplan.cpp:8357]: It was introduced at MySQL and for MySQL. So I attributed it to MySQL.

File: dbcon/mysql/ha_mcs_execplan.cpp (hunk 8 above)
Comment (mariadb-SergeyZefirov) [reply dbcon/mysql/ha_mcs_execplan.cpp:8411]: Yes.

File: dbcon/execplan/calpontselectexecutionplan.cpp (hunk 1 above)
Comment (mariadb-SergeyZefirov) [reply dbcon/execplan/calpontselectexecutionplan.cpp:108]: Yes. It is explained in the comment next to groupByKeysCount() functions of CSEP.

File: dbcon/mysql/ha_mcs_execplan.cpp (hunk 8 above)
Comment (drrtuy) [reply dbcon/mysql/ha_mcs_execplan.cpp:8411]: Where exactly?

File: dbcon/mysql/ha_mcs_execplan.cpp (hunk 7 above)
Comment (drrtuy) [reply dbcon/mysql/ha_mcs_execplan.cpp:8357]: This project is not MySQL anymore and moreover a real effort has been done on renaming binaries so plz change it to MariaDB to follow the common strategy.

File: dbcon/execplan/calpontselectexecutionplan.cpp (hunk 1 above)
Comment (drrtuy) [reply dbcon/execplan/calpontselectexecutionplan.cpp:108]: Makes sense.

**Comment by mariadb-SergeyZefirov** [comment]
I close this PR because I moved development to the  main repository.

The new PR is this one: https://github.com/mariadb-corporation/mariadb-columnstore-engine/pull/3194

**Pull request closed without merging**